domain/application/service/charm.go

```diff
@@ -163,8 +163,8 @@ func (s *CharmService) GetCharmID(ctx context.Context, args charm.GetCharmArgs)
 		return "", applicationerrors.CharmNameNotValid
 	}
 
-	if rev := args.Revision; rev != nil && *rev >= 0 {
-		return s.st.GetCharmIDByRevision(ctx, args.Name, *rev)
+	if rev := args.Revision; rev != nil && *rev >= 0 && args.Source != "" {
+		return s.st.GetCharmID(ctx, args.Name, *rev, args.Source)
```

Comment:
It should be flat-out error if we don't know the source. Can we validate the source exists and is either:

 - local
 - charmhub (ch).

Reply:
Oh sure! d'uh :rofl: I was thinking about this use when writing the parse method